<algorithm>: fill does not accept volatile byte pointers

**Describe the bug**
`fill` cannot be called using a `volatile byte*`, as it triggers an optimized call to `memset`, which does not accept `volatile` pointers. This bug does not manifest with `char` types. See related bug #1183 

**Command-line test case**
```
E:\test>type main.cpp
#include <algorithm>
#include <cstddef>

template <class T>
void volatile_zero(void* ptr, std::size_t size)
{
    auto begin = static_cast<volatile T*>(ptr);
    constexpr T val{ 0 };
    std::fill_n(begin, size, val);
}

int main()
{
    int data;
    volatile_zero<char>(&data, sizeof(data));
    volatile_zero<signed char>(&data, sizeof(data));
    volatile_zero<unsigned char>(&data, sizeof(data));
    volatile_zero<std::byte>(&data, sizeof(data));
}

E:\test>cl /EHsc /std:c++17 .\main.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29617 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

main.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29617\include\xutility(4833): error C2664: 'void *memset(void *,int,size_t)': cannot convert argument 1 from '_DestTy *const ' to 'void *'
        with
        [
            _DestTy=std::byte
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29617\include\xutility(4833): note: Conversion loses qualifiers
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29617\include\vcruntime_string.h(63): note: see declaration of 'memset'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29617\include\xutility(4930): note: see reference to function template instantiation 'void std::_Fill_memset<volatile T,_Ty>(_DestTy *const ,const _Ty,const size_t)' being compiled
        with
        [
            T=std::byte,
            _Ty=std::byte,
            _DestTy=std::byte
        ]
.\main.cpp(9): note: see reference to function template instantiation '_OutIt std::fill_n<volatile T*,size_t,T>(_OutIt,const _Diff,const _Ty &)' being compiled
        with
        [
            _OutIt=volatile std::byte *,
            T=std::byte,
            _Diff=size_t,
            _Ty=std::byte
        ]
.\main.cpp(18): note: see reference to function template instantiation 'void volatile_zero<std::byte>(void *,size_t)' being compiled
```

**Expected behavior**
I would expect this to compile.

**STL version**
Microsoft Visual Studio Community 2019 Preview
Version 16.9.0 Preview 2.0

--- a/stl/inc/xutility
+++ b/stl/inc/xutility
@@ -4352,10 +4352,11 @@ template <>
 struct _Is_character_or_byte_or_bool<bool> : true_type {};
 
 // _Fill_memset_is_safe determines if _FwdIt and _Ty are eligible for memset optimization in fill
+// Need to explicity test for volatile because _Unwrap_enum_t discards qualifiers.
 template <class _FwdIt, class _Ty, bool = is_pointer_v<_FwdIt>>
 _INLINE_VAR constexpr bool _Fill_memset_is_safe = conjunction_v<is_scalar<_Ty>,
     _Is_character_or_byte_or_bool<_Unwrap_enum_t<remove_reference_t<_Iter_ref_t<_FwdIt>>>>,
-    is_assignable<_Iter_ref_t<_FwdIt>, const _Ty&>>;
+    negation<is_volatile<remove_reference_t<_Iter_ref_t<_FwdIt>>>>, is_assignable<_Iter_ref_t<_FwdIt>, const _Ty&>>;
 
 template <class _FwdIt, class _Ty>
 _INLINE_VAR constexpr bool _Fill_memset_is_safe<_FwdIt, _Ty, false> = false;

--- a/stl/inc/xutility
+++ b/stl/inc/xutility
@@ -4351,8 +4351,8 @@ struct _Is_character_or_byte_or_bool<byte> : true_type {};
 template <>
 struct _Is_character_or_byte_or_bool<bool> : true_type {};
 
-// _Fill_memset_is_safe determines if _FwdIt and _Ty are eligible for memset optimization in fill
-// Need to explicity test for volatile because _Unwrap_enum_t discards qualifiers.
+// _Fill_memset_is_safe determines if _FwdIt and _Ty are eligible for memset optimization in fill.
+// Need to explicitly test for volatile because _Unwrap_enum_t discards qualifiers.
 template <class _FwdIt, class _Ty, bool = is_pointer_v<_FwdIt>>
 _INLINE_VAR constexpr bool _Fill_memset_is_safe = conjunction_v<is_scalar<_Ty>,
     _Is_character_or_byte_or_bool<_Unwrap_enum_t<remove_reference_t<_Iter_ref_t<_FwdIt>>>>,

--- a/tests/std/tests/VSO_0180469_fill_family/test.cpp
+++ b/tests/std/tests/VSO_0180469_fill_family/test.cpp
@@ -134,6 +134,9 @@ int main() {
     test_fill<char, int>();
 
     test_fill<volatile char, char>(); // Test GH-1183
+#ifdef __cpp_lib_byte
+    test_fill<volatile byte, byte>(); // Test GH-1556
+#endif // __cpp_lib_byte
 
     test_uninitialized_fill(
         [](count_copies* buff, size_t n, const count_copies& src) { uninitialized_fill(buff, buff + n, src); });

